### PR TITLE
Merge consequent FilterSteps.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow merging two consequent `FilterSteps` of a query plan. This improves filter-push-down optimization if the filter condition can be pushed down from the parent step.

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit e449f83cfaa548cec8b17ddfec1123e0a7b27c84 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64760/e449f83cfaa548cec8b17ddfec1123e0a7b27c84/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp [hunk 1]
```diff
@@ -45,6 +46,10 @@ void optimizePrewhere(Stack & stack, QueryPlan::Nodes &)
     if (!source_step_with_filter)
         return;
 
+    /// Prewhere is broken with StorageMerge and old analyzer.
+    if (!source_step_with_filter->getContext()->getSettingsRef().allow_experimental_analyzer && typeid_cast<ReadFromMerge *>(source_step_with_filter))
```
Comment (vdimir) [inline src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp:50]: For example this one I've been investigated is related https://github.com/ClickHouse/ClickHouse/issues/63105 
But what are other issues?

**Review by vdimir — APPROVED** [review]

File: src/Storages/StorageMerge.cpp [hunk 2]
```diff
@@ -1228,7 +1228,7 @@ QueryPlan ReadFromMerge::createPlanForTable(
 
         if (allow_experimental_analyzer)
         {
-            InterpreterSelectQueryAnalyzer interpreter(modified_query_info.query_tree,
+            InterpreterSelectQueryAnalyzer interpreter(modified_query_info.query_tree->toAST(),
```
Comment (vdimir) [inline src/Storages/StorageMerge.cpp:1231]: Why need to convert to ast?

File: src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp (hunk 1 above)
Comment (vdimir) [reply src/Processors/QueryPlan/Optimizations/optimizePrewhere.cpp:50]: That issue I'm trying to fix by recalculating child plans after prewhere info is updated https://github.com/ClickHouse/ClickHouse/pull/64893

**Pull request merged**